### ISSUE TITLE
Custom id value

### ISSUE THREAD

**Issue by egalitarian** [josiahcarlson/rom#59]
Setting custom value for id (e.g. like guid) should be supported.

**Comment by josiahcarlson** [comment]
Pretend for a moment that what is currently considered to be the "id" column was invisible, but that you also had a `.guid` column. If this guid column were auto-generating by default, but you could set to something explicitly if you chose, and you could get the object by the value, prefix, and suffix, would that suffice?

```
import uuid
import rom

class GuidPrimaryKeyExample(rom.Model):
    _id = rom.PrimaryKey()
    guid = rom.String(default=lambda:str(uuid.uuid4),
        unique=True, prefix=True, suffix=True)
```

Add columns and methods as necessary, and that's it. Now, there are *very* good reasons to use an integer primary key, even if it is hidden, among which being that if you have any indexes, your memory utilization can be as much as 1/4 as the size of a guid primary key. But this gets you the guid you want, with the integer primary key that rom needs to be efficient.

I think that's enough support. Do you agree?

**Comment by egalitarian** [comment]
I agree with you about the memory foorprint of guid as primary key. There are still use cases where you have to set your own integer id for the exact same reason - to reduce the memory overhead of having hidden and custom id. Somethimes you have calculated/hashed ids or even ids imported from another database that needs to be kept. By not allowing to set the id to custom value if its integer we need to keep two ids. Also there's the added problem that an integer field can not be set to be unique which a second id should really be.

**Comment by josiahcarlson** [comment]
An auto-generating integer primary key in rom is not going away. There are several assumptions that various parts of rom make and rely upon for correct operation, and removing/replacing that functionality is more work than the potential benefits of reducing memory by 1 column per entity. Case in point: the default index overhead (when you have no indexes on any column) is 2-3x as large as storing the integer primary key, so the potential space savings are negligible in practice.

Rather than trying to request the support for a different kind of primary key, or for self-defined primary keys, please realize that the existing rom primary key is an underlying implementation detail that you can ignore. And the other indexes you can add to other columns should offer the majority of modeling/indexing options that are reasonable*.

* That integer columns in rom can't currently be used as a unique key is definitely a missing feature. I can't remember why I made that decision when I first built rom, but at this point it's not important. I've added support for unique integer columns with a short test case, and will cut a release later today if you don't have any other feature requests.

**Comment by egalitarian** [comment]
I understand your reasoning. Unique integer column is a good enough approximation to the functionality I was looking for. Thanks

**Comment by josiahcarlson** [comment]
I just tagged and released version 0.29.5 that offers unique index support on integer columns. It should be available in PyPI and in the repository.
